test/unittests/UserTest.php

```diff
@@ -0,0 +1,526 @@
+    }
+
+    /**
+     * Tears down the fixture, for example, close a network connection.
+     * This method is called after a test is executed.
+     *
+     * @return void
+     */
+    protected function tearDown()
+    {
+        parent::tearDown();
+        $this->_factory->reset();
+    }
+
+    /**
+     * Test factory() method retrieves all _candidate and related info
+     *
+     * @return void
+     * @covers User::singleton
+     * @covers User::factory
+     * @covers User::getData
+     */
+    public function testFactoryRetrievesUserInfo()
+    {
+        $this->_setUpTestDoublesForFactoryUser();
+        $this->_user = \User::factory($this->_username);
+        //validate _user Info
+        $this->assertEquals($this->_userInfoComplete, $this->_user->getData());
+    }
+
+    /**
+     * Test that getFullname returns the correct full name of the user
+     *
+     * @return void
+     * @covers User::getFullname
+     */
+    public function testGetFullname()
+    {
+        $this->_user = \User::factory($this->_username);
+        $this->assertEquals(
+            $this->_userInfoComplete['Real_name'],        
+            $this->_user->getFullname()
+        );
+    }
+
+    /**
+     * Test that getId returns the correct ID of the user
+     *
+     * @return void
+     * @covers User::getId
+     */
+    public function testGetId()
+    {
+        $this->_user = \User::factory($this->_username);
+        $this->assertEquals(
+            $this->_userInfoComplete['ID'],
+            $this->_user->getId()
+        );
+    }
+
+    /**
+     * Test that getUsername returns the correct UserID of the user
+     *
+     * @return void
+     * @covers User::getUsername
+     */
+    public function testGetUsername()
+    {
+        $this->_user = \User::factory($this->_username);
+        $this->assertEquals(
+            $this->_userInfoComplete['UserID'],
+            $this->_user->getUsername()
+        );
+    }
+
+    /**
+     * Test that getSiteNames returns the site names in the correct format
+     * TODO Add more sites
+     *
+     * @return void
+     * @covers User::getSiteNames
+     */
+    public function testSiteNames()
+    {
+        $this->_user = \User::factory($this->_username);
+        $this->assertEquals(
+            array('psc_test', 'psc_test2'),
+            $this->_user->getSiteNames()
+        );
+    }
+
+    /**
+     * Test that getCenterIDs returns the correct array of center IDs of the user
+     *
+     * @return void
+     * @covers User::getCenterIDs
+     */
+    public function testGetCenterIDs()
+    {
+        $this->_user = \User::factory($this->_username);
+        $this->assertEquals(
+            $this->_userInfoComplete['CenterIDs'],
+            $this->_user->getCenterIDs()
+        );
+    }
+
+    /**
+     * Test that getLanguagePreference returns the correct integer from the user
+     * TODO getLanguagePreference is returning a string rather than an int. 
```

Comment:
The function signature in the User class looks like getLanguagePreference() can only return `int`. What error are you getting with this test?

Reply:
The error is that it is returning a string rather than an int. It's strange because in the "users" table, the language_preferences row is supposed to be an int, but when you use either getData('language_preference') or getLanguagePreference(), it returns a string. I think when the user object gets populated from the database (in the factory() method), it saves it as a string rather than an int!

Reply:
I can edit the factory() method so that it saves 'language_preference' as an int, or I can change the getLanguagePreference method so that it expects a string. Which do you think is best @johnsaigle ?

Reply:
Anything that's selected from the database comes back as a string unless explicitly cast. That's just the way the database driver works in PHP.

Reply:
So should I just change getLanguagePreference so that it expects a string?

Reply:
I'm not really familiar with getLanguagePreference, but if the parameter is really an integer encoded as a string I think it would be better to cast it to an int before calling than to use a string representation of a number.

Reply:
I think you can update the function itself to be `return intval(...)` and that should fix it.

Reply:
OK so in the factory method, I can just cast it to an int when it's setting up the user object.

Reply:
Yeah that should work.